model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto

```diff
@@ -438,42 +441,77 @@ message ProcessBundleSplitRequest {
     // possible and returning the remainder).
     double fraction_of_remainder = 1;
 
-    // A set of allowed element indices where the SDK may split. When this is
-    // empty, there are no constraints on where to split.
+    // (Optional) A set of allowed element indices where the SDK may split. When
+    // this is empty, there are no constraints on where to split.
     repeated int64 allowed_split_points = 3;
 
-    // (Required for GrpcRead operations) Number of total elements expected
-    // to be sent to this GrpcRead operation, required to correctly account
-    // for unreceived data when determining where to split.
+    // (Required for gRPC Read operation transforms) Number of total elements
+    // expected to be sent to this GrpcRead operation, required to correctly
+    // account for unreceived data when determining where to split.
     int64 estimated_input_elements = 2;
-
-    // TODO(SDF): Allow providing weights rather than sizes.
   }
 
   // (Required) Specifies the desired split for each transform.
   //
-  // Currently only splits at GRPC read operations are supported.
+  // Currently only splits at gRPC read operations are supported.
   // This may, of course, limit the amount of work downstream operations
   // receive.
   map<string, DesiredSplit> desired_splits = 3;
 }
 
-// Represents a partition of the bundle: a "primary" and
-// a "residual", with the following properties:
+// Represents a partition of the bundle: a "primary" and a "residual", with the
+// following properties:
 // - The work in primary and residual doesn't overlap, and combined, adds up
 //   to the work in the current bundle if the split hadn't happened.
 // - The current bundle, if it keeps executing, will have done none of the
-//   work under residual_roots.
+//   work under residual_roots and none of the elements in the
 // - The current bundle, if no further splits happen, will have done exactly
-//   the work under primary_roots.
+//   the work under primary_roots and all elements up to and including the
+//   channel splits last_primary_element.
+//
```

Comment:
It might be valuable to have a conceptual summary here, e.g. "This allows the SDK to relinquish ownership of and commit to not process some of the elements that it may have been sent (the residual) while retaining ownership and commitment to finish the other portion (the primary)."